File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/IsProductActive/ProductActiveStatusRepository.cs
using CreditCardEnrollment.Common.Projection;
using MongoDB.Driver;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.IsProductActive;

public class ProductActiveStatusRepository {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;
    private static string _collectionName = "CreditCard_Enrollment_ProductActiveStatus";

    public ProductActiveStatusRepository(MongoTransactionalProjectionOperator mongoOperator) {
        _mongoOperator = mongoOperator;
    }

    public bool IsThereAnActiveProductWithId(string productId) {
        return _mongoOperator.Find<ProductActiveStatus>(
            _collectionName,
            p => p.Id == productId && p.Active
        ).Any();
    }

    public ProductActiveStatus? FindOneById(string productId) {
        return _mongoOperator.Find<ProductActiveStatus>(
            _collectionName,
            p => p.Id == productId
        ).FirstOrDefault();
    }

    public void Save(ProductActiveStatus productActiveStatus) {
        _mongoOperator.ReplaceOne(
            _collectionName,
            p => p.Id == productActiveStatus.Id, 
            productActiveStatus, 
            new ReplaceOptions { IsUpsert = true }
        );
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/projection/isProductActive/ProductActiveStatusRepository.ts
import {ProductActiveStatus} from "./ProductActiveStatus";
import {MongoTransactionalProjectionOperator} from "../../../../common/projection/MongoTransactionalProjectionOperator";
import {inject, injectable} from "tsyringe";

@injectable()
export class ProductActiveStatusRepository {
    private readonly collectionName = 'CreditCard_Enrollment_ProductActiveStatus';

    constructor(
        @inject(MongoTransactionalProjectionOperator) private readonly mongoOperator: MongoTransactionalProjectionOperator
    ) {}

    async isThereAnActiveProductWithId(productId: string): Promise<boolean> {
        const results = await this.mongoOperator.find<ProductActiveStatus>(
            this.collectionName,
            { _id: productId, active: true }
        );
        return results.length > 0;
    }

    async findOneById(productId: string): Promise<ProductActiveStatus | null> {
        const results = await this.mongoOperator.find<ProductActiveStatus>(this.collectionName, { _id: productId });
        return results.length > 0 ? results[0] : null;
    }

    async save(productActiveStatus: ProductActiveStatus): Promise<void> {
        await this.mongoOperator.replaceOne(
            this.collectionName,
            { _id: productActiveStatus._id },
            productActiveStatus,
            { upsert: true }
        );
    }
}